nevergrad/functions/powersystems/core.py

```diff
@@ -0,0 +1,154 @@
+                stocks[i] += 0.5*(1.+np.cos(2*pi*t/(24*365) + delay[i])) * np.random.rand()
+            # Consumption model.
+            base_consumption = (self.constant_to_year_ratio*self.year_to_day_ratio 
+                    +0.5*self.year_to_day_ratio*(1.+np.cos(2*pi*t/(24*365))) + 0.5*(1.+np.cos(2*pi*t/24)))
+            if t == 0:
+                consumption = base_consumption
+            else:
+                consumption = max(0., consumption + self.consumption_noise*(np.random.rand()-.5) + self.back_to_normal * (base_consumption - consumption))
+            consumption_per_time_step += [consumption]
+            # Water usage.
+            needed = consumption
+    
+            # Setting inputs for all agents.
+            base_x = [np.cos(2*pi*t/24.), np.sin(2*pi*t/24.), np.cos(2*pi*t/(365*24)), np.sin(2*pi*t/(365*24)), needed, self.average_consumption, self.year_to_day_ratio, self.constant_to_year_ratio, self.back_to_normal, self.consumption_noise]
+            x = list(base_x + self.thermal_power_capacity + self.thermal_power_prices + stocks)
+    
+            # Prices as a decomposition tool!
+            price: List[float] = [a.GetOutput(np.array(x))[0][0] for a in dam_managers]
+            volume: List[float] = [s for s in stocks]
+            dam_index: List[int] = list(range(len(price)))
+            price += self.thermal_power_prices
+            volume += self.thermal_power_capacity
+            dam_index += [-1] * len(price)
+            
+            assert(len(price) == N + self.num_thermal_plants)
+            hydro_prod: List[float] = [0.] * N
+            for i in range(len(price)):
+                if needed == 0:
+                    break
+                production = min(volume[i], needed)
+                if dam_index[i] >= 0:
+                    hydro_prod[dam_index[i]] += production
+                    stocks[dam_index[i]] -= production
+                    assert(stocks[dam_index[i]] >= -1e-7)
+                else:
+                    cost += production * price[i]
+                needed -= production
+            # Cost in case of failures -- this is
+            # harming industries and hospitals.
+            cost += 500. * needed
+            hydro_prod_per_time_step += hydro_prod
+        return cost  # Other data of interest: , hydro_prod, hydro_prod_per_time_step, consumption_per_time_step
+
+    def __init__(self, num_stocks: int = 13, depth: int = 3, width: int = 3) -> None:
+        # Number of stocks (dams).
+        self.N = num_stocks
+        N = self.N
+        # Parameters describing the problem.
+        self.year_to_day_ratio = 2.  # Ratio between variation of consumption in the year and variation of consumption in the day
+        self.constant_to_year_ratio = 1.
+        self.back_to_normal = 0.5  # How much of the gap with normal is cancelled at each iteration.
+        self.consumption_noise = 0.1
+        self.num_thermal_plants = 7
+        self.number_of_years = 1
```

Comment:
These parameters are defined in infra.py. They should be defined only once to make sure they are the same everywhere. If you keep them here,  it feels like most of them should be parameters of the init function instead of hardcoded constants (you may want to change the horizon or number of plants).

Reply:
Done.